### PR TITLE
[New feature] Migrate old caggs to new form

### LINKED ISSUE #1519: [New feature] Migration from new to old caggs
TimescaleDB 2.8 includes a feature to help migrate old (partialized) caggs to the new (formalized) format. Add instructions for the migration.

https://github.com/timescale/timescaledb/pull/4552

### PR DESCRIPTION
**Pull request by charislam**
Document procedure, added in TimescaleDB 2.8, that lets you migrate old
caggs (pre-2.7 format) to the new format.

# Links

Fixes #1519 

# Writing help

For information about style and word usage, see the [style guide](https://docs.timescale.com/timescaledb/latest/contribute-to-docs/)

# Review checklists
Reviewers: use this section to ensure you have checked everything before approving this PR:

## Subject matter expert (SME) review checklist

- [ ] Is the content technically accurate?
- [ ] Is the content complete?
- [ ] Is the content presented in a logical order?
- [ ] Does the content use appropriate names for features and products?
- [ ] Does the content provide relevant links to further information?

## Documentation team review checklist

- [x] Is the content free from typos?
- [x] Does the content use plain English?
- [x] Does the content contain clear sections for concepts, tasks, and references?
- [x] Have any images been uploaded to the correct location, and are resolvable?
- [x] If the page index was updated, are redirects required
      and have they been implemented?
- [ ] Have you checked the built version of this content?

### CODE REVIEW

File: timescaledb/how-to-guides/continuous-aggregates/migrate.md
Comment (Loquacity) [inline timescaledb/how-to-guides/continuous-aggregates/migrate.md:15]:
```suggestion
use this command. Connect to your database and run:
```

It's not really a procedure?

File: timescaledb/how-to-guides/continuous-aggregates/index.md
Comment (fabriziomello) [inline timescaledb/how-to-guides/continuous-aggregates/index.md:24]:
```suggestion
*   [Migrate your continuous aggregates][cagg-migrate] from old to new format.
```

File: timescaledb/how-to-guides/continuous-aggregates/migrate.md
Comment (fabriziomello) [reply timescaledb/how-to-guides/continuous-aggregates/migrate.md:15]: Yes, it is a procedure. Maybe we can write `use the provided Postgres procedure.` WDYT?

**Comment by fabriziomello** [comment]
@charislam What about add information about how to check the migration progress/status??? It is possible executing a SELECT query on the `_timescaledb_catalog.continuous_agg_migrate_plan_step`

File: timescaledb/how-to-guides/continuous-aggregates/migrate.md
Comment (Loquacity) [reply timescaledb/how-to-guides/continuous-aggregates/migrate.md:15]: Ah, I see what you mean. A procedure in the Postgres sense, not the documentation sense 😅 

```suggestion
use this procedure. Connect to your database and run:
```

File: timescaledb/how-to-guides/continuous-aggregates/migrate.md
Comment (charislam) [inline timescaledb/how-to-guides/continuous-aggregates/migrate.md:15]:
```suggestion
use this procedure. It automatically copies over your data and policies.
```

File: timescaledb/how-to-guides/continuous-aggregates/migrate.md
Comment (Loquacity) [inline timescaledb/how-to-guides/continuous-aggregates/migrate.md:25]: I think Boolean needs initcaps?

File: timescaledb/how-to-guides/continuous-aggregates/index.md
Comment (jfjoly) [inline timescaledb/how-to-guides/continuous-aggregates/index.md:24]: Should we be more specific and say `old (pre TimescaleDB 2.7)` ?

File: timescaledb/how-to-guides/continuous-aggregates/migrate.md
Comment (jfjoly) [inline timescaledb/how-to-guides/continuous-aggregates/migrate.md:16]: We could add here that the procedure is transparent and users can continue to use the continuous aggregate during the migration.

File: timescaledb/how-to-guides/continuous-aggregates/index.md
Comment (charislam) [reply timescaledb/how-to-guides/continuous-aggregates/index.md:24]:
```suggestion
*   [Migrate your continuous aggregates][cagg-migrate] from old to new format.
    Continuous aggregates created in TimescaleDB 2.7 and above are in the new
    format, unless explicitly created in the old format.
```

File: timescaledb/how-to-guides/continuous-aggregates/migrate.md
Comment (charislam) [inline timescaledb/how-to-guides/continuous-aggregates/migrate.md:15]:
```suggestion
To migrate a continuous aggregate from the old format to the new format, you can
use this procedure. It automatically copies over your data and policies. You can
continue to use the continuous aggregate while the migration is happening.
```

File: timescaledb/how-to-guides/continuous-aggregates/migrate.md
Comment (charislam) [reply timescaledb/how-to-guides/continuous-aggregates/migrate.md:25]: Not for the abstract data type, according to Google:

> In most contexts, you're talking about a specific data type in a specific programming language, in which case use code font and the exact spelling and capitalization of the programming keyword. But when referring to the abstract data type, use lowercase. (In the unlikely event that you refer to Boolean mathematics or Boolean logic, use uppercase.)

**Review by fabriziomello — APPROVED** [review]

File: timescaledb/how-to-guides/continuous-aggregates/migrate.md
Comment (Loquacity) [reply timescaledb/how-to-guides/continuous-aggregates/migrate.md:26]: Who knew?! 🤷‍♀️

**Review by Loquacity — APPROVED** [review]

**Comment by charislam** [comment]
Direct links for easy previewing:

http://docs-dev.timescale.com/docs-charis-migrate-caggs/timescaledb/charis-migrate-caggs/how-to-guides/continuous-aggregates/migrate/
http://docs-dev.timescale.com/docs-charis-migrate-caggs/api/charis-migrate-caggs/continuous-aggregates/cagg_migrate/

**Comment by jfjoly** [comment]
Some of the arguments , starting with underscores, are wrong: 
_cagg
_override
_drop_old

Not sure if that gets replaced automatically. The link to the preview pages are not working for me because of a certificate issue. 

This is the UX:
```
-- DEFAULT OPTION: migrate to the new CAGG and rename the old one
CALL cagg_migrate_to_finals_form ('cagg_name');

-- Migrate to the new CAGG and delete the old one
CALL cagg_migrate_to_finals_form ('cagg_name', drop_old_cagg => TRUE);

-- Migrate to the new CAGG with a new name and keep the existing one
CALL cagg_migrate_to_finals_form ('cagg_name', swap_at_end => FALSE);
```

**Comment by Loquacity** [comment]
> The link to the preview pages are not working for me because of a certificate issue.

Try opening the previews in incognito mode.

**Comment by jfjoly** [comment]
Please hold on publishing this as we found an issue. 
https://github.com/timescale/timescaledb/issues/4682

**Comment by fabriziomello** [comment]
> Some of the arguments , starting with underscores, are wrong: _cagg _override _drop_old
> 
> Not sure if that gets replaced automatically. The link to the preview pages are not working for me because of a certificate issue.
> 
> This is the UX:
> 
> ```
> -- DEFAULT OPTION: migrate to the new CAGG and rename the old one
> CALL cagg_migrate_to_finals_form ('cagg_name');
> 
> -- Migrate to the new CAGG and delete the old one
> CALL cagg_migrate_to_finals_form ('cagg_name', drop_old_cagg => TRUE);
> 
> -- Migrate to the new CAGG with a new name and keep the existing one
> CALL cagg_migrate_to_finals_form ('cagg_name', swap_at_end => FALSE);
> ```

Actually @jfjoly the parameters starts with underscore

**Comment by jfjoly** [comment]
Fabrizio also recommended to add some warning in the docs about the permissions or catalog tables that store the cagg migration state...

I would suggest to add it on the how to guide at the bottom 
For instance:

> **Configure permissions (only required if you get permissions error when executing the migration)**
> The user that execute the procedure should have INSERT/UPDATE grants in the following tables (in the case the user is not a superuser, of course):
> `GRANT UPDATE ON _timescaledb_catalog.continuous_agg_migrate_plan TO %user%;`
> `GRANT UPDATE ON _timescaledb_catalog.continuous_agg_migrate_plan_step TO %user%;`
>

**Comment by fabriziomello** [comment]
> Fabrizio also recommended to add some warning in the docs about the permissions or catalog tables that store the cagg migration state...
> 
> I would suggest to add it on the how to guide at the bottom For instance:
> 
> > **Configure permissions (only required if you get permissions error when executing the migration)**
> > The user that execute the procedure should have INSERT/UPDATE grants in the following tables (in the case the user is not a superuser, of course):
> > `GRANT UPDATE ON _timescaledb_catalog.continuous_agg_migrate_plan TO %user%;`
> > `GRANT UPDATE ON _timescaledb_catalog.continuous_agg_migrate_plan_step TO %user%;`

Just a quick fix:

**Configure permissions (only required if you get permissions error when executing the migration)**
The user that execute the procedure should have SELECT/INSERT/UPDATE grants in the following objects (in the case the user is not a superuser, of course):
```sql
GRANT SELECT, INSERT, UPDATE ON TABLE _timescaledb_catalog.continuous_agg_migrate_plan TO %user%;
GRANT SELECT, INSERT, UPDATE ON TABLE _timescaledb_catalog.continuous_agg_migrate_plan_step TO %user%;
GRANT USAGE ON SEQUENCE _timescaledb_catalog.continuous_agg_migrate_plan_step_step_id_seq TO %user%;
```
Where `%user%` should be replaced for the real user name.

**Comment by charislam** [comment]
@fabriziomello @jfjoly Documented! Is it good to merge with this fix?

**Comment by fabriziomello** [comment]
> @fabriziomello @jfjoly Documented! Is it good to merge with this fix?

LGTM

**Comment by iroussos** [comment]
Following the discussion in [slack](https://iobeam.slack.com/archives/C032AS21X4P/p1662822106158919) between @jfjoly and @fabriziomello, there are two different issues to address:

1. The permissions required to perform the migration
2. The parameter names

I agree with @jfjoly on holding any communication, updating the parameter names and then releasing the docs.

@fabriziomello do you mind adding the PR to change the parameter names? 

@charislam We can then also update the parameter names in this PR and include both changes in the next minor release. 

Our only risk is someone using the new undocumented function and hitting the errors. Do you agree with this plan @jfjoly?

**Comment by fabriziomello** [comment]
@charislam There are an ongoing PR to fix the `cagg_migrate` parameter names (https://github.com/timescale/timescaledb/pull/4694). I'll let you know once it get merged.

**Comment by fabriziomello** [comment]
> @charislam There are an ongoing PR to fix the `cagg_migrate` parameter names ([timescale/timescaledb#4694](https://github.com/timescale/timescaledb/pull/4694)). I'll let you know once it get merged.

Merged. /cc @jfjoly

**Comment by iroussos** [comment]
I have a few newbie questions - sorry for the noise but I am still new and trying to understand our processes:

1. @charislam Am I right that docs updates go live the moment we merge them?

    If so, other than updating the parameter names in this PR to follow the fix in https://github.com/timescale/timescaledb/pull/4694, we'll have to wait and merge this PR once the next patch is released. /cc @jfjoly to correct me on this one

2. @charislam Do we somehow mark functions (or functionality) in the docs with the version they are introduced?

    I may missed this, but I don't see any reference to 2.8 in `cagg_migrate`, how are users going to know which version of TimescaleDB they should be running to be able to use the functionality? We mention 2.7 (when the new caggs format was introduced) but not the version the function became available.  
    
    Similarly, how are we going to mark it as 2.8.1+ as the version in 2.8.0 will not work with the params in the documentation?

3. @fabriziomello how's the process of doing a patch release? Is it time consuming, do we try to bundle multiple fixes or do we go wild on patch releases?

    On this specific case, should we start the process for 2.8.1 or is it too early?
    
Thank you!

**Comment by charislam** [comment]
> I have a few newbie questions - sorry for the noise but I am still new and trying to understand our processes:
> 
>     1. @charislam Am I right that docs updates go live the moment we merge them?
>        If so, other than updating the parameter names in this PR to follow the fix in [Change parameter names of cagg_migrate procedure timescaledb#4694](https://github.com/timescale/timescaledb/pull/4694), we'll have to wait and merge this PR once the next patch is released. /cc @jfjoly to correct me on this one
> 
>     2. @charislam Do we somehow mark functions (or functionality) in the docs with the version they are introduced?
>        I may missed this, but I don't see any reference to 2.8 in `cagg_migrate`, how are users going to know which version of TimescaleDB they should be running to be able to use the functionality? We mention 2.7 (when the new caggs format was introduced) but not the version the function became available.
>        Similarly, how are we going to mark it as 2.8.1+ as the version in 2.8.0 will not work with the params in the documentation?
> 
>     3. @fabriziomello how's the process of doing a patch release? Is it time consuming, do we try to bundle multiple fixes or do we go wild on patch releases?
>        On this specific case, should we start the process for 2.8.1 or is it too early?
> 
> 
> Thank you!

@iroussos Those are very, very good questions!

1. Yes, they go live immediately, so if the patch is not out yet, we should wait for it to go out.
2. We do not have versioning currently in the docs. Adding 'introduced in version' labels is on the backlog, for exactly this reason! The first rendition will probably just be a label on the entire API reference. Figuring out what to do with renamed parameters is a more complex question and we'll have to do some thinking about how we display that info for readers given we don't have full-scale versioning

File: api/cagg_migrate.md
Comment (iroussos) [inline api/cagg_migrate.md:42]:
```suggestion
|`cagg`|`REGCLASS`|The continuous aggregate to migrate|
```

File: api/cagg_migrate.md
Comment (iroussos) [inline api/cagg_migrate.md:48]:
```suggestion
|`override`|`BOOLEAN`|If false, the old continuous aggregate keeps its name. The new continuous aggregate is named `<OLD_CONTINUOUS_AGGREGATE_NAME>_new`. If true, the new continuous aggregate gets the old name. The old continuous aggregate is renamed `<OLD_CONTINUOUS_AGGREGATE_NAME>_old`. Defaults to `false`.|
```

File: api/cagg_migrate.md
Comment (iroussos) [inline api/cagg_migrate.md:49]:
```suggestion
|`drop_old`|`BOOLEAN`|If true, the old continuous aggregate is deleted. Must be used together with `override`. Defaults to `false`.|
```

File: timescaledb/how-to-guides/continuous-aggregates/migrate.md
Comment (iroussos) [inline timescaledb/how-to-guides/continuous-aggregates/migrate.md:37]:
```suggestion
`override` and `drop_old`. By default, the name of your new continuous
aggregate is the name of your old continuous aggregate, with the suffix `_new`.

Set `override` to true to rename your new continuous aggregate with the
original name. The old continuous aggregate is renamed with the suffix `_old`.

To both rename and drop the old continuous aggregate entirely, set both
parameters to true. Note that `drop_old` must be used together with
`override`.
```

**Review by iroussos — CHANGES_REQUESTED** [review]
@charislam I have a few proposed changes to update the parameter names in this PR to follow the fix in https://github.com/timescale/timescaledb/pull/4694

Do you mind checking if everything is right and updating the PR?

We plan to move forward with the release of 2.8.1 (https://github.com/timescale/eng-database/issues/321), so it would be great to have the docs PR ready to be merged

**Comment by charislam** [comment]
> @charislam I have a few proposed changes to update the parameter names in this PR to follow the fix in [timescale/timescaledb#4694](https://github.com/timescale/timescaledb/pull/4694)
> 
> Do you mind checking if everything is right and updating the PR?
> 
> We plan to move forward with the release of 2.8.1 ([timescale/eng-database#321](https://github.com/timescale/eng-database/issues/321)), so it would be great to have the docs PR ready to be merged

@iroussos everything looks good! I committed your suggested changes and solved the merge conflicts, so we should be ready to merge whenever you give the 👍🏼

**Comment by iroussos** [comment]
Thank you @charislam, much appreciated 🚀

**Comment by charislam** [comment]
@iroussos I saw that 2.8.1 is tagged, am I good to merge this now?

**Comment by iroussos** [comment]
@charislam I agree with @fabriziomello, we'll ping you once 2.8.1 is publicly available to our cloud users

**Review by iroussos — APPROVED** [review]
@charislam on our side we are ready to merge the doc changes; 2.8.1 has been released to MST and will be released as well to Timescale cloud today.

/cc @jnidzwetzki as the release captain

**Pull request merged**
